configuration/exodus-pipeline.yaml

```diff
@@ -46,6 +50,10 @@ Conditions:
     !Equals [!Ref env, stage]
   IsProd:
     !Equals [!Ref env, prod]
+  IsExodus:
```

Comment:
I'm not really a fan of this part, since part of the reason to provision this with a different project in the first place would be testing the pipeline, but forcibly disabling chunks of it whenever project isn't named "exodus" makes it impossible to test those parts.

Is it possible to have an explicit parameter instead, named after what it does?  e.g. useCloudtrails with default of false?

Reply:
The problem is that, once created as part of a CloudFormation stack, the CloudTrail trail will be locked to that stack.
So, if a dev creates a stack to test the pipeline before the real pipelines are created, the dev would have to delete their stack, or at least break it by deleting the trail, before the real pipelines can go up.

With having `IsExodusProd`, we ensure that the trail is tied to the real environment, and devs can then use the trail in their own pipelines afterward. The only thing you'd be unable to test post-creation is the creation iteself.

Reply:
> The only thing you'd be unable to test post-creation is the creation itself.

+1.  In theory, if our account uses a single trail to monitor all s3 write API calls, it should be a very rare occurrence for us to ever modify our existing trail, thus reducing the need to test the creation of the CloudTrail resource.

Though, if strongly preferred, is possible for us to maintain separate trails. I can prepend the ${project} var to each resource so they can be created in separate stacks. We will just have to be aware that we only have room for two development CloudTrail trails, and as such, only two developers can test their stage -> prod transitions in parallel. Thus, we will need to strictly enforce that each developer only create their CloudTrail resources immediately before testing, and delete their CloudTrail resources immediately after testing. This could be accomplished by deploying the developer's prod pipeline with the "useCloudtrails=true" param before testing, and then without the "useCloudtrails" param after testing. The only real negative of this solution is a limited number of resources, and the strict requirement of enforcing clean-up. Maintaining separate trails would also enable us to reduce the data we collect (reducing the monitoring to only one {project}-pipeline-artifacts bucket per trail), as Nathan suggested in an earlier comment. 

Note that the reduction in data collection is only useful if we do not want to collect data for s3 write API calls in our production exodus AWS account. If we want to collect data for all s3 write API calls for security/auditing reasons, I think we should stick with the single trail and accept that the creation of a shared, account-wide resource is not something a developer will not need to test.

Reply:
I suppose it makes some sense that one may want to test the prod pipeline by itself, without promotion from stage. So, I'm okay with having a CloudTrail parameter if we're set on multiple trails.

However, I don't like the idea of steering devs into the AWS console/cli and delete individual resources -- seems like it could get messy. We should probably try to limit interaction with resources to CloudFormation create/update, delete. I.e., if one enables CloudTrail with their pipeline stack, the trail should be deleted with the their pipeline stack. I'm not sure if something like that's enforceable, though.

Reply:
If it matters, I'm in favor of the single trail with a log expiration of 1 day. We may have other uses for CloudTrail in the future.

Reply:
When writing my earlier comments, I didn't really understand how the cloudtrail was connected to our other resources.  Now taking the time to try to understand it, based on the EventSelectors it seems that one trail is going to monitor events for resources provisioned from *all* values of ${project} and ${env} so there really is no point to permanently maintain more than one.

I might not have objected to IsExodusProd if I understood that at the time.  Nevertheless, I like the way it is now which does seem to support temporarily provisioning a new cloudtrail if needed.  A reason why it might be needed is if we want to try changing the settings (e.g. the EventSelector) on the trail.  It would be better to create a separate one and try out the proposed config there rather than immediately deploying the changes to the single production trail.